oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticBulkProcessorHandler.java

```diff
@@ -0,0 +1,304 @@
+            // register new bulk party
+            phaser.register();
+
+            // init update status
+            updatesMap.put(executionId, Boolean.FALSE);
+
+            LOG.debug("Sending bulk with id {} -> {}", executionId, bulkRequest.getDescription());
+            if (LOG.isTraceEnabled()) {
+                LOG.trace("Bulk Requests: \n{}", bulkRequest.requests()
+                        .stream()
+                        .map(DocWriteRequest::toString)
+                        .collect(Collectors.joining("\n"))
+                );
+            }
+        }
+
+        @Override
+        public void afterBulk(long executionId, BulkRequest bulkRequest, BulkResponse bulkResponse) {
+            LOG.debug("Bulk with id {} processed with status {} in {}", executionId, bulkResponse.status(), bulkResponse.getTook());
+            if (LOG.isTraceEnabled()) {
+                try {
+                    LOG.trace(Strings.toString(bulkResponse.toXContent(jsonBuilder(), EMPTY_PARAMS)));
+                } catch (IOException e) {
+                    LOG.error("Error decoding bulk response", e);
+                }
+            }
+            if (bulkResponse.hasFailures()) { // check if some operations failed to execute
+                Set<String> failedDocSet = new LinkedHashSet<>();
+                NodeBuilder status = definitionBuilder.child(IndexDefinition.STATUS_NODE);
+                // Read the current failed paths (if any) on the :status node into failedDocList
+                if (status.hasProperty(IndexDefinition.FAILED_DOC_PATHS)) {
+                    for (String str : status.getProperty(IndexDefinition.FAILED_DOC_PATHS).getValue(Type.STRINGS)) {
+                        failedDocSet.add(str);
+                    }
+                }
+
+                int initialSize = failedDocSet.size();
+                boolean isFailedDocSetFull = false;
+
+                for (BulkItemResponse bulkItemResponse : bulkResponse) {
+                    if (bulkItemResponse.isFailed()) {
+                        BulkItemResponse.Failure failure = bulkItemResponse.getFailure();
+                        if (!isFailedDocSetFull && failedDocSet.size() < FAILED_DOC_COUNT_FOR_STATUS_NODE) {
+                            failedDocSet.add(bulkItemResponse.getId());
+                        } else {
+                            isFailedDocSetFull = true;
+                        }
+                        // Log entry to be used to parse logs to get the failed doc id/path if needed
+                        LOG.error("ElasticIndex Update Doc Failure: Error while adding/updating doc with id : [{}]", bulkItemResponse.getId());
+                        LOG.error("Failure Details: BulkItem ID: " + failure.getId() + ", Failure Cause: {}", failure.getCause());
+                    } else {
+                        // Set indexUpdated to true even if 1 item was updated successfully
+                        updatesMap.put(executionId, Boolean.TRUE);
```

Comment:
It will repeatedly update the map. I don't think it's an issue, because ConcurrentHashMaps are fast, but I guess it's not intended.

Reply:
Good catch. Although it wasn't a huge problem since this code block should be executed frequently (only when there are failures) I have introduced a flag to avoid updating the map multiple times.